sdk/storage/azure-storage-datamovement/src/job_engine_task_generation.cpp

```diff
@@ -49,6 +51,108 @@ namespace Azure { namespace Storage { namespace _detail {
+        auto description = Core::Json::_internal::json::array();
+        for (const auto& range : subtaskRanges)
+        {
+          description.push_back(range.Offset);
+          description.push_back(range.Length.Value());
+        }
+        subtaskRanges.clear();
+        if (!description.empty())
+        {
+          subtasksDescription.push_back(std::move(description));
+        }
+        totalDownloadSize += subtaskDownloadSize;
+        subtaskDownloadSize = 0;
+      };
+      std::function<void(const Azure::Core::Http::HttpRange&)> addRange;
+      addRange = [&](const Azure::Core::Http::HttpRange& range) {
+        constexpr int64_t MergeThreshold = 512;
+        AZURE_ASSERT(range.Length.Value() > 0);
+        int64_t distance = subtaskRanges.empty()
+            ? std::numeric_limits<int64_t>::max()
+            : range.Offset - (subtaskRanges.back().Offset + subtaskRanges.back().Length.Value());
+
+        if (range.Length.Value() > task.ChunkSize)
+        {
+          if (distance > 0 && distance <= MergeThreshold
+              && task.ChunkSize - subtaskDownloadSize > distance)
+          {
+            Azure::Core::Http::HttpRange newRange;
+            newRange.Offset = range.Offset - distance;
+            newRange.Length = distance + range.Length.Value();
+            addRange(newRange);
+          }
+          else if (task.ChunkSize - subtaskDownloadSize > 0)
+          {
+            Azure::Core::Http::HttpRange newRange;
+            newRange.Offset = range.Offset;
+            newRange.Length = task.ChunkSize - subtaskDownloadSize;
+            addRange(newRange);
+            Azure::Core::Http::HttpRange newRange2;
+            newRange2.Offset = range.Offset + newRange.Length.Value();
+            newRange2.Length = range.Length.Value() - newRange.Length.Value();
+            addRange(newRange2);
+          }
+          else
+          {
+            flushSubtask();
+            addRange(range);
+          }
+        }
+        else
+        { // range.Length.Value() <= task.ChunkSize
+          if (distance <= MergeThreshold
+              && subtaskDownloadSize + distance + range.Length.Value() <= task.ChunkSize)
+          {
+            subtaskRanges.back().Length.Value() += distance + range.Length.Value();
+            subtaskDownloadSize += distance + range.Length.Value();
+          }
+          else if (
+              distance > MergeThreshold
+              && subtaskDownloadSize + range.Length.Value() <= task.ChunkSize)
+          {
+            subtaskRanges.push_back(range);
+            subtaskDownloadSize += range.Length.Value();
+          }
+          else
+          {
+            flushSubtask();
+            addRange(range);
+          }
+        }
+      };
+
+      for (auto rangePage = pageBlobClient.GetPageRanges(); rangePage.HasPage();
+           rangePage.MoveToNextPage())
+      {
+        for (const auto& range : rangePage.PageRanges)
+        {
+          addRange(range);
```

Comment:
I didn't go through all range handling code. If you have done this, please just ignore this comment.
It might be slow to send one download request for each small range. It'd be faster to merge nearby ranges to a larger one (maybe 1MB  or so) and download a larger chunk with one request.

Reply:
I defined a constant variable at Line 80. The threshold is now 512 bytes, but it can be easily changed to any value.
`constexpr int64_t MergeThreshold = 512;`

`1MB` seems too large. The overhead of issuing a new HTTP request is: http request header size (around 700\~800 bytes), http response header size (around 700\~800 bytes), one extra transaction, one extra round-trip time. With these numbers in mind, I think 512 bytes or 1024 bytes might be a better choice.